Fix axeDev tools + lighthouse error for learner dashboard and make the lighthouse score to 1

Patel-Muhammad opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes the axe devtools errors and lighthouse errors for learner dashboard
2. This PR does the following: Adds aria-label, img alt text and keyboard support for the feedback table in learner dashboard

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

https://github.com/oppia/oppia/assets/96219910/020ff941-32ae-4fd2-9432-bdce3d6ee069


https://github.com/oppia/oppia/assets/96219910/90cede7f-a29d-4b4f-8289-8ee45d9209f9



<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
![Screenshot 2023-06-03 210236](https://github.com/oppia/oppia/assets/96219910/c8c63b5e-a867-48cc-afdc-39a5c1b967cf)

#### Lighthouse testscore for desktop
![Screenshot 2023-06-03 205458](https://github.com/oppia/oppia/assets/96219910/e559ed05-cadf-48b4-b2d5-6d067274e91f)


#### Axe devtools 
![Screenshot 2023-06-03 210519](https://github.com/oppia/oppia/assets/96219910/876ca826-d912-47e0-8b0f-dca92cce9d6f)
![Screenshot 2023-06-03 210506](https://github.com/oppia/oppia/assets/96219910/2b45b417-2d74-45d2-b000-8d8dd3b8a4d6)
(The 5 issues in this screenshot are related to the color contrast of the exploration tiles that is fixed in #18328 )

![Screenshot 2023-06-03 210453](https://github.com/oppia/oppia/assets/96219910/c0c8c24a-a245-409c-a6b7-5c8dbcda1a8c)
![Screenshot 2023-06-03 210439](https://github.com/oppia/oppia/assets/96219910/1ff6ed08-ea38-4dc6-a35c-a84ae5175908)
![Screenshot 2023-06-03 210426](https://github.com/oppia/oppia/assets/96219910/2ad0bb14-d253-4cfa-bbfb-77170ffd4b03)



<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Patel-Muhammad committed:
Added a11y fixes

Patel-Muhammad committed:
Fixed lint

Patel-Muhammad committed:
Lint

Patel-Muhammad committed:
Added alt text

Patel-Muhammad committed:
testing

Patel-Muhammad committed:
added keyboard support for feedback

Patel-Muhammad committed:
fixed lint

Patel-Muhammad committed:
changed the keyboard support for feedback

Patel-Muhammad committed:
fixed typo

Patel-Muhammad committed:
alt

Patel-Muhammad committed:
added more voice support

Rijuta-s commented on core/templates/components/summary-tile/learner-topic-goals-summary-tile.component.html:
```diff
@@ -3,7 +3,7 @@
       <div class="story-card">
         <div class="oppia-title-section" [ngStyle]="{'background-color': thumbnailBgColor}">
-          <img class="oppia-thumbnail-image" [src]="thumbnailUrl" [attr.style]="isCardHovered()">
+          <img alt="Thumbnail image for the story node {{storyNodeTitle}}" class="oppia-thumbnail-image" [src]="thumbnailUrl" [attr.style]="isCardHovered()">
```
```suggestion
          <img alt="Thumbnail image for the story card {{storyNodeTitle}}" class="oppia-thumbnail-image" [src]="thumbnailUrl" [attr.style]="isCardHovered()">
```
Alternate text should be very simple and understandable, I think we should use 'card' instead of node.

Patel-Muhammad replied:
Done

Patel-Muhammad committed:
fixed lint

Rijuta-s commented on core/templates/pages/learner-dashboard-page/progress-tab.component.html:
```diff
@@ -54,8 +54,8 @@
           </span>
-          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown image" class="oppia-svg-dropdown-image" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
-          <img [src]="getStaticImageUrl('/learner_dashboard/up.svg')" alt="up image" class="oppia-svg-up-image" (click)="showSkills(i)" *ngIf="displaySkills[i]">
+          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown image" class="oppia-svg-dropdown-image" role="button" tabindex="0" (click)="showSkills(i)" (keydown.enter)="showSkills(i)" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
```
```suggestion
          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown button" class="oppia-svg-dropdown-image" role="button" tabindex="0" (click)="showSkills(i)" (keydown.enter)="showSkills(i)" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
```
It is serving as a button, so the alt text should also indicate that it is a button, rather than a image, ditto elsewhere.

Patel-Muhammad replied:
Done

Rijuta-s commented on core/templates/pages/learner-dashboard-page/progress-tab.component.html:
```diff
@@ -54,8 +54,8 @@
-          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown image" class="oppia-svg-dropdown-image" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
-          <img [src]="getStaticImageUrl('/learner_dashboard/up.svg')" alt="up image" class="oppia-svg-up-image" (click)="showSkills(i)" *ngIf="displaySkills[i]">
+          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown image" class="oppia-svg-dropdown-image" role="button" tabindex="0" (click)="showSkills(i)" (keydown.enter)="showSkills(i)" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
+          <img [src]="getStaticImageUrl('/learner_dashboard/up.svg')" alt="up image" class="oppia-svg-up-image" role="button" tabindex="0" (click)="showSkills(i)" (keydown.enter)="showSkills(i)" (click)="showSkills(i)" *ngIf="displaySkills[i]">
```
```suggestion
          <img [src]="getStaticImageUrl('/learner_dashboard/up.svg')" alt="up button" class="oppia-svg-up-image" role="button" tabindex="0" (click)="showSkills(i)" (keydown.enter)="showSkills(i)" (click)="showSkills(i)" *ngIf="displaySkills[i]">
```

Patel-Muhammad replied:
Done

Patel-Muhammad committed:
Merge branch 'develop' into learner-dashboard-1

Rijuta-s commented:
@Patel-Muhammad In the video shown above, in the 'Goals' tab, it did not read out the completed goals section?

Patel-Muhammad committed:
Update core/templates/components/summary-tile/learner-topic-goals-summary-tile.component.html

Co-authored-by: Rijuta Singh <68547101+Rijuta-s@users.noreply.github.com>

Patel-Muhammad committed:
changes

Patel-Muhammad commented:
@Rijuta-s this video shows completed goals section is being readout:



https://github.com/oppia/oppia/assets/96219910/90e7ba98-f0e8-488e-b3f1-ae1c7a6fd26e

Rijuta-s commented:
@Patel-Muhammad Reply to the comments with 'done' once you have addressed them for future.

Patel-Muhammad commented:
@Rijuta-s I have done that, PTAL. Thanks.

Rijuta-s reviewed (changes requested):
Blocking to make sure discussed changes are done.

Patel-Muhammad committed:
revised

Patel-Muhammad commented:
Hi @Rijuta-s  I have addressed the requested changes, PTAL.


- **Please note:** When "send"  button is clicked in the feedback thread it stays in "sending" state, even in oppia.org and on testserver

https://github.com/oppia/oppia/assets/96219910/c04bfc61-0911-4588-b800-e3cc8699e076
https://github.com/oppia/oppia/assets/96219910/2f7847d2-a839-4f3c-9476-ef42ab3456a6



- Other than that: I have added the `aria-live="polite"`

https://github.com/oppia/oppia/assets/96219910/2ad18459-ca4b-41e9-8369-fa44bbd5fd2d



- There are 5 issues in axe that are related to color contrast of the oppia lesson tile 
- About the cursor moving: When I have to stop the recording I have to click the stop button, that is why in the end of the video the cursor appears to move.

Please let me know if there are some other queries.
Thanks.

Patel-Muhammad committed:
time

Patel-Muhammad committed:
Merge branch 'develop' into learner-dashboard-1

Rijuta-s commented on core/templates/pages/learner-dashboard-page/progress-tab.component.html:
```diff
@@ -1,9 +1,9 @@
 <div class="progress-tab">
-  <div class="skill-proficiency" [ngClass]="{'empty-section': emptySkillProficiency}" [ngStyle]="{'margin-top': windowIsNarrow? '35px': '64px'}" *ngIf="!windowIsNarrow || (windowIsNarrow && activeSubsection === LEARNER_DASHBOARD_SUBSECTION_I18N_IDS.SKILL_PROFICIENCY)">
+  <div tabindex="0" class="skill-proficiency" [ngClass]="{'empty-section': emptySkillProficiency}" [ngStyle]="{'margin-top': windowIsNarrow? '35px': '64px'}" *ngIf="!windowIsNarrow || (windowIsNarrow && activeSubsection === LEARNER_DASHBOARD_SUBSECTION_I18N_IDS.SKILL_PROFICIENCY)">
```
```suggestion
  <div tabindex="0"
       class="skill-proficiency"
       [ngClass]="{'empty-section': emptySkillProficiency}"
       [ngStyle]="{'margin-top': windowIsNarrow? '35px': '64px'}"
       *ngIf="!windowIsNarrow || (windowIsNarrow && activeSubsection === LEARNER_DASHBOARD_SUBSECTION_I18N_IDS.SKILL_PROFICIENCY)">
```
this makes html files easily readible. Ditto elsewhere if the sentence is more than 80 chars

Patel-Muhammad replied:
Done

Rijuta-s commented:
Thanks @Patel-Muhammad
1) hmm I see, a bug needs to be filled for the send button then, it should change to sent, once the backend reports success.
2) In the video at 0:33, why does on opening feedback thread focus shifts back to 'feedback updates' button? I don't think that should happen. Also at around 0:35, the voiceover is - 'return to list of messages' and after a brief pause it says 'button'. This is very confusing.
3) Also is it possible for voice reader to only read 'sending'  not 'sending ellipses'? If not, its okay.
4) Color contrast issues are to be fixed in another PR right? Can you please mention this below the video proof itself?

Rijuta-s commented on core/templates/pages/learner-dashboard-page/progress-tab.component.html:
```diff
@@ -54,8 +54,8 @@
           </span>
-          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown image" class="oppia-svg-dropdown-image" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
-          <img [src]="getStaticImageUrl('/learner_dashboard/up.svg')" alt="up image" class="oppia-svg-up-image" (click)="showSkills(i)" *ngIf="displaySkills[i]">
+          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown button" class="oppia-svg-dropdown-image" role="button" tabindex="0" (keydown.enter)="showSkills(i)" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
```
```suggestion
          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')"
               alt="dropdown button"
               class="oppia-svg-dropdown-image"
               role="button" tabindex="0"
               (keydown.enter)="showSkills(i)"
               (click)="showSkills(i)"
               *ngIf="!displaySkills[i]">
```

Patel-Muhammad replied:
Done

Rijuta-s commented on core/templates/pages/learner-dashboard-page/progress-tab.component.html:
```diff
@@ -54,8 +54,8 @@
-          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown image" class="oppia-svg-dropdown-image" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
-          <img [src]="getStaticImageUrl('/learner_dashboard/up.svg')" alt="up image" class="oppia-svg-up-image" (click)="showSkills(i)" *ngIf="displaySkills[i]">
+          <img [src]="getStaticImageUrl('/learner_dashboard/dropdown.svg')" alt="dropdown button" class="oppia-svg-dropdown-image" role="button" tabindex="0" (keydown.enter)="showSkills(i)" (click)="showSkills(i)" *ngIf="!displaySkills[i]">
+          <img [src]="getStaticImageUrl('/learner_dashboard/up.svg')" alt="up button" class="oppia-svg-up-image" role="button" tabindex="0" (keydown.enter)="showSkills(i)" (click)="showSkills(i)" *ngIf="displaySkills[i]">
```
Ditto as above

Patel-Muhammad replied:
Done

Rijuta-s commented on core/templates/components/summary-tile/learner-topic-summary-tile.component.html:
```diff
@@ -3,7 +3,7 @@
      class="oppia-topic-link">
     <div class="oppia-title-section" [ngStyle]="{'background-color': thumbnailBgColor}">
-      <img class="oppia-thumbnail-image" [src]="thumbnailUrl">
+      <img alt="" class="oppia-thumbnail-image" [src]="thumbnailUrl">
```
I will suggest to add a comment where ever `alt=""`  is being used. The comment should outline the importance of empty alt text being added. This will ensure that developers in future just don't simply remove it thinking that it serve no purpose.

Patel-Muhammad replied:
Done

Patel-Muhammad committed:
Update core/templates/pages/learner-dashboard-page/progress-tab.component.html

Co-authored-by: Rijuta Singh <68547101+Rijuta-s@users.noreply.github.com>

Patel-Muhammad committed:
Update core/templates/pages/learner-dashboard-page/progress-tab.component.html

Co-authored-by: Rijuta Singh <68547101+Rijuta-s@users.noreply.github.com>

Patel-Muhammad committed:
Added focus on the feedback and comments

Patel-Muhammad committed:
Merge branch 'develop' into learner-dashboard-1

Rijuta-s commented on core/templates/components/summary-tile/learner-story-summary-tile.component.html:
```diff
@@ -6,7 +6,8 @@
       <div class="story-card">
         <div class="oppia-title-section" [ngStyle]="{'background-color': thumbnailBgColor}">
-          <img class="oppia-thumbnail-image" [src]="thumbnailUrl" [attr.style]="isCardHovered()">
+          // The alt text is empty as the image is just a thumbnail image, it is not supposed to convey any important information to the user
```
```suggestion
          // We add an empty alt text so that the screen reader does not read the thumbnail image filename and ignores it as the thumbnail does not convey any significant information.
```

Patel-Muhammad replied:
Done

Patel-Muhammad commented:
Hi @Rijuta-s, I have made the requested changes:
1. Added comments
2. The focus is now on the feedback thread when it open
3. Now it reads: "Return to the list of messages button"  and then reads button
4. I have filed the issue in #18352 

### Video of the changes:
https://github.com/oppia/oppia/assets/96219910/243698c5-427b-4945-b294-2ae4554d7fff

Patel-Muhammad committed:
Update core/templates/components/summary-tile/learner-story-summary-tile.component.html

Co-authored-by: Rijuta Singh <68547101+Rijuta-s@users.noreply.github.com>

Patel-Muhammad committed:
changed the comment

Rijuta-s commented:
@Patel-Muhammad In the video it does not go to the 'return to the list of message button'. Also the thing I wanted to bring focus on was that after reading 'return to the list of message button', it again reads button which seems confusing as if the focus has shifted from the previous button to some other button

Patel-Muhammad committed:
fixed import

Patel-Muhammad commented:
@Rijuta-s it reads "button" in the end, that is intentional and done for the clarity by default. As it says to the user what type of element is in focus.

Patel-Muhammad committed:
lint

Patel-Muhammad committed:
Merge branch 'develop' into learner-dashboard-1

Patel-Muhammad commented:
@Rijuta-s please take a look at the video showing the focus going to return button:

https://github.com/oppia/oppia/assets/96219910/82a37677-31dd-44dd-b61b-d7a271842dde

Patel-Muhammad committed:
spaces

Patel-Muhammad committed:
initialised exptitleref

Patel-Muhammad committed:
changed the logic

Patel-Muhammad committed:
Merge branch 'develop' into learner-dashboard-1

Patel-Muhammad committed:
lint

Patel-Muhammad committed:
test

Patel-Muhammad committed:
test

Rijuta-s commented on core/templates/pages/learner-dashboard-page/goals-tab.component.html:
```diff
@@ -59,12 +69,12 @@
       <img [src]="bookImageUrl" alt="book image" class="oppia-svg-book-image" *ngIf="editGoalsTopicClassification[i] == 1" ngbTooltip="You have completed this topic.">
-      <mat-checkbox color="primary" *ngIf="editGoalsTopicClassification[i] == 2" class="oppia-checkbox e2e-test-add-topic-to-current-goals-button" (click)="addToLearnerGoals(topic, topic.id, i)" [disabled]="this.currentGoals.length == this.MAX_CURRENT_GOALS_LENGTH"></mat-checkbox>
-      <mat-checkbox color="primary" *ngIf="editGoalsTopicClassification[i] == 0" checked disabled class="oppia-checkbox"></mat-checkbox>
+      <mat-checkbox aria-label="{{ topic.name }} checkbox" color="primary" *ngIf="editGoalsTopicClassification[i] == 2" class="oppia-checkbox e2e-test-add-topic-to-current-goals-button" (click)="addToLearnerGoals(topic, topic.id, i)" [disabled]="this.currentGoals.length == this.MAX_CURRENT_GOALS_LENGTH"></mat-checkbox>
```
```suggestion
      <mat-checkbox aria-label="{{ topic.name }}
                    checkbox"
                    color="primary"
                    *ngIf="editGoalsTopicClassification[i] == 2"
                    class="oppia-checkbox e2e-test-add-topic-to-current-goals-button
                    (click)="addToLearnerGoals(topic, topic.id, i)"
                    [disabled]="this.currentGoals.length == this.MAX_CURRENT_GOALS_LENGTH">
       </mat-checkbox>
```

Patel-Muhammad replied:
Done

Rijuta-s commented on core/templates/pages/learner-dashboard-page/goals-tab.component.html:
```diff
@@ -59,12 +69,12 @@
-      <mat-checkbox color="primary" *ngIf="editGoalsTopicClassification[i] == 2" class="oppia-checkbox e2e-test-add-topic-to-current-goals-button" (click)="addToLearnerGoals(topic, topic.id, i)" [disabled]="this.currentGoals.length == this.MAX_CURRENT_GOALS_LENGTH"></mat-checkbox>
-      <mat-checkbox color="primary" *ngIf="editGoalsTopicClassification[i] == 0" checked disabled class="oppia-checkbox"></mat-checkbox>
+      <mat-checkbox aria-label="{{ topic.name }} checkbox" color="primary" *ngIf="editGoalsTopicClassification[i] == 2" class="oppia-checkbox e2e-test-add-topic-to-current-goals-button" (click)="addToLearnerGoals(topic, topic.id, i)" [disabled]="this.currentGoals.length == this.MAX_CURRENT_GOALS_LENGTH"></mat-checkbox>
+      <mat-checkbox aria-label="{{ topic.name }} checkbox is disabled" color="primary" *ngIf="editGoalsTopicClassification[i] == 0" checked disabled class="oppia-checkbox"></mat-checkbox>
```
```suggestion
      <mat-checkbox checked disabled
                    aria-label="{{ topic.name }}
                    checkbox is disabled"
                    color="primary"
                    *ngIf="editGoalsTopicClassification[i] == 0"
                    class="oppia-checkbox">
     </mat-checkbox>
```

Patel-Muhammad replied:
Done

Rijuta-s reviewed (approved):
Just 2 minor comments, otherwise looks good to me! Thanks!!

Patel-Muhammad committed:
changed the focus logic

Patel-Muhammad committed:
lint

Patel-Muhammad committed:
added test

Patel-Muhammad committed:
lint

Patel-Muhammad committed:
test

Patel-Muhammad committed:
lint

Patel-Muhammad committed:
Update core/templates/pages/learner-dashboard-page/goals-tab.component.html

Co-authored-by: Rijuta Singh <68547101+Rijuta-s@users.noreply.github.com>

Patel-Muhammad committed:
Update core/templates/pages/learner-dashboard-page/goals-tab.component.html

Co-authored-by: Rijuta Singh <68547101+Rijuta-s@users.noreply.github.com>

Patel-Muhammad committed:
Merge branch 'develop' into learner-dashboard-1

Patel-Muhammad committed:
lint

Patel-Muhammad committed:
lint

Patel-Muhammad committed:
removed branch name in const file

Patel-Muhammad committed:
chnaged the lint

Patel-Muhammad commented:
Hi @seanlip  this PR passes all the checks and is also approved by the reviewers. Can you please merge it? Thanks!

seanlip commented:
No, it hasn't been approved by all the reviewers. See the part at the bottom for whom you're missing:

![Screenshot from 2023-06-10 23-12-09](https://github.com/oppia/oppia/assets/10575562/900523be-ff13-4fa2-a2ff-4ef7de2b2bcb)

Patel-Muhammad commented:
Hii @kevintab95 @JayVivarekar, can you please review this PR? Thanks.

JayVivarekar reviewed (approved):
Thanks @Patel-Muhammad! LGTM

Patel-Muhammad committed:
Merge branch 'develop' into learner-dashboard-1

Patel-Muhammad commented:
Hi @seanlip this PR is approved by the reviewers and passes all the tests, PTAL. Thanks!
